File: frontend/src/domain/auto-speed.ts
export interface AutoSpeedRule {
  type: "sustained" | "burst";
  threshold_mbps: number;
  sustained_seconds: number;
  window_seconds: number;
  burst_count: number;
  limit_mbps: number;
  limit_duration: number;
}

export function validAutoSpeedRule(rule: AutoSpeedRule): boolean {
  const rate = (value: number) => Number.isFinite(value) && value * 125000 >= 1 && value * 125000 <= 2 ** 50;
  const integer = (value: number, min: number, max: number) => Number.isInteger(value) && value >= min && value <= max;
  return (rule.type === "sustained" || rule.type === "burst") && rate(rule.threshold_mbps) && rate(rule.limit_mbps)
    && integer(rule.sustained_seconds, 1, 86400) && integer(rule.limit_duration, 1, 86400)
    && integer(rule.window_seconds, 0, 86400) && integer(rule.burst_count, 0, 10000)
    && (rule.type !== "burst" || (rule.window_seconds >= rule.sustained_seconds && rule.burst_count >= 1));
}

export function newAutoSpeedRule(): AutoSpeedRule {
  return { type: "sustained", threshold_mbps: 50, sustained_seconds: 30,
    window_seconds: 300, burst_count: 3, limit_mbps: 10, limit_duration: 60 };
}

File: backend/app/open_node/domain/auto_speed.py
"""Validated automatic speed rules shared by plan and native policy APIs."""

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator


class AutoSpeedRule(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)

    type: Literal["sustained", "burst"]
    threshold_mbps: float = Field(ge=1 / 125000, le=(1 << 50) / 125000, allow_inf_nan=False)
    sustained_seconds: int = Field(ge=1, le=86400)
    window_seconds: int = Field(default=0, ge=0, le=86400)
    burst_count: int = Field(default=0, ge=0, le=10000)
    limit_mbps: float = Field(ge=1 / 125000, le=(1 << 50) / 125000, allow_inf_nan=False)
    limit_duration: int = Field(ge=1, le=86400)

    @model_validator(mode="after")
    def valid_burst(self):
        if self.type == "burst" and (
            self.window_seconds < self.sustained_seconds or self.burst_count < 1
        ):
            raise ValueError("Burst rules require a valid window and occurrence count")
        return self
